File: src/main/java/pl/mjedynak/idea/plugins/pit/configuration/PitConfigurationType.java
package pl.mjedynak.idea.plugins.pit.configuration;

import com.intellij.execution.configuration.ConfigurationFactoryEx;
import com.intellij.execution.configurations.ConfigurationFactory;
import com.intellij.execution.configurations.ConfigurationType;
import com.intellij.execution.configurations.ModuleBasedConfiguration;
import com.intellij.execution.configurations.RunConfiguration;
import com.intellij.openapi.extensions.Extensions;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.IconLoader;
import com.intellij.util.containers.ContainerUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.Icon;

public class PitConfigurationType implements ConfigurationType {

    private static final Icon ICON = IconLoader.getIcon("/pit.gif");
    private static final String DISPLAY_NAME = "PIT Runner";
    private static final String ID = "PIT";
    private static final String CONFIGURATION_DESCRIPTION = "Executes PIT mutation testing";

    private final ConfigurationFactory myFactory;

    public PitConfigurationType() {
        myFactory = new ConfigurationFactoryEx(this) {
            public RunConfiguration createTemplateConfiguration(Project project) {
                PitRunConfigurationFactory pitRunConfigurationFactory = new PitRunConfigurationFactory();
                return pitRunConfigurationFactory.createConfiguration(project);
            }

            @Override
            public Icon getIcon(@NotNull final RunConfiguration configuration) {
                return getIcon();
            }

            @Override
            public void onNewConfigurationCreated(@NotNull RunConfiguration configuration) {
                ((ModuleBasedConfiguration) configuration).onNewConfigurationCreated();
            }
        };
    }

    @Override
    public String getDisplayName() {
        return DISPLAY_NAME;
    }

    @Override
    public String getConfigurationTypeDescription() {
        return CONFIGURATION_DESCRIPTION;
    }

    @Override
    public Icon getIcon() {
        return ICON;
    }

    @Override
    public ConfigurationFactory[] getConfigurationFactories() {
        return new ConfigurationFactory[]{myFactory};
    }

    @Override
    @NotNull
    public String getId() {
        return ID;
    }

    @Nullable
    public static PitConfigurationType getInstance() {
        return ContainerUtil.findInstance(Extensions.getExtensions(CONFIGURATION_TYPE_EP), PitConfigurationType.class);
    }
}

File: src/main/kotlin/pl/mjedynak/idea/plugins/pit/configuration/PitConfigurationType.kt
package pl.mjedynak.idea.plugins.pit.configuration

import com.intellij.execution.configurations.ConfigurationFactory
import com.intellij.execution.configurations.ConfigurationType
import com.intellij.execution.configurations.RunConfiguration
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.IconLoader
import com.intellij.util.containers.ContainerUtil
import org.jetbrains.annotations.NonNls
import javax.swing.Icon

class PitConfigurationType : ConfigurationType {
    private val myFactory: ConfigurationFactory =
        object : ConfigurationFactory(this) {
            override fun createTemplateConfiguration(project: Project): RunConfiguration {
                val pitRunConfigurationFactory = PitRunConfigurationFactory()
                return pitRunConfigurationFactory.createConfiguration(project)
            }

            override fun getIcon(configuration: RunConfiguration): Icon = this@PitConfigurationType.icon

            override fun getId(): @NonNls String = name
        }

    override fun getDisplayName(): String = DISPLAY_NAME

    override fun getConfigurationTypeDescription(): String = CONFIGURATION_DESCRIPTION

    override fun getIcon(): Icon = ICON

    override fun getConfigurationFactories(): Array<ConfigurationFactory> = arrayOf(myFactory)

    override fun getId(): @NonNls String = ID

    companion object {
        private val ICON: Icon = IconLoader.getIcon("/pit.svg", PitConfigurationType::class.java)
        private const val DISPLAY_NAME = "PIT Runner"
        private const val ID = "PIT"
        private const val CONFIGURATION_DESCRIPTION = "Executes PIT mutation testing"

        fun getInstance(): PitConfigurationType? =
            ContainerUtil.findInstance(
                ConfigurationType.CONFIGURATION_TYPE_EP.extensionList,
                PitConfigurationType::class.java,
            )
    }
}
